cluster-autoscaler/expander/priority/priority.go

```diff
@@ -0,0 +1,78 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package priority
+
+import (
+	"strconv"
+
+	"k8s.io/autoscaler/cluster-autoscaler/expander"
+	"k8s.io/autoscaler/cluster-autoscaler/expander/random"
+	"k8s.io/klog"
+	schedulernodeinfo "k8s.io/kubernetes/pkg/scheduler/nodeinfo"
+)
+
+const (
+	scalingPriorityLabel = "zalando.org/scaling-priority"
+)
+
+type priority struct {
+	fallback expander.Strategy
+}
+
+// NewStrategy returns a scale up strategy (expander) that picks the node group with the highest priority.
+func NewStrategy() expander.Strategy {
+	return &priority{
+		fallback: random.NewStrategy(),
+	}
+}
+
+// BestOption selects the expansion option based on the highest priority nodes
+func (ps *priority) BestOption(options []expander.Option, nodeInfo map[string]*schedulernodeinfo.NodeInfo) *expander.Option {
+	priorityOptions := map[int][]expander.Option{}
```

Comment:
We don't really need to build the map, since we're only interested in the highest priority pool (and it makes the code more complex). You can keep the list of options for the currently highest priority, and then either append to it if the next option has the same priority, skip the next option if its priority is lower, or reset the list to just the next one if the priority is higher.

Reply:
> You can also avoid the issue of pools with the sentinel priority being ignored by checking if the list is empty and then resetting the priority.

Which case do you mean exactly?

Reply:
Nevermind that, I thought that it'd behave incorrectly if the only pool in the cluster was with a negative priority (and this is why you started those at 0), but this isn't actually the case.